modules/popup/react/stories/stories.tsx

```diff
@@ -79,3 +80,110 @@ export const RTL = () => {
     </CanvasProvider>
   );
 };
+
+export const Accessible = () => {
+  const {targetProps, closePopup, popperProps, stackRef} = usePopup();
+
+  const isElementNativelyFocusable = (element: HTMLElement) => {
+    const nodeName = element.nodeName.toLowerCase();
+    const validInput = nodeName === 'input' && (element as HTMLInputElement).type !== 'hidden';
+    const validAnchor = nodeName === 'a' && element.hasAttribute('href');
+    const validAudioVideo =
+      ['audio', 'video'].includes(nodeName) && element.hasAttribute('controls');
+    const validImgObject = ['img', 'object'].includes(nodeName) && element.hasAttribute('usemap');
+    const validNativelyFocusable = [
+      'button',
+      'details',
+      'embed',
+      'iframe',
+      'select',
+      'textarea',
+    ].includes(nodeName);
+
+    return validInput || validAnchor || validAudioVideo || validImgObject || validNativelyFocusable;
+  };
+
+  const getFirstFocusableElement = (content: HTMLElement): HTMLElement | null => {
+    const elements = content.querySelectorAll('*');
+
+    for (let i = 0; i < elements.length; i++) {
+      const element = elements.item(i);
+      if (element && isElementNativelyFocusable(element as HTMLElement)) {
+        return element as HTMLElement;
+      }
+    }
+
+    console.log(
+      'No focusable element was found. Please ensure popup has at least one focusable element'
+    );
+    return null;
+  };
+
+  const useInitialFocus = (popupRef: React.RefObject<HTMLElement>, popupOpen: boolean = false) => {
+    React.useLayoutEffect(() => {
+      const handlerRef =
+        document.activeElement instanceof HTMLElement ? document.activeElement : null;
+
+      if (popupRef.current && popupOpen) {
+        const elem = getFirstFocusableElement(popupRef.current);
+        if (elem) {
+          elem.focus();
+        }
+      }
+      return () => {
+        if (handlerRef) {
+          handlerRef.focus();
+        }
+      };
```

Comment:
This isn't exactly what we want. This callback will be called when any dependencies have changed. Since `popupRef` will never change, that leaves `popupOpen` which changes both when the popup is opened and when it closes. It happens to work in our simple example, but what actually happens when the `popupOpen` goes from `false` to `true`:

- `handlerRef` is focused
- A new layout effect callback is set up and `elem.focus` is called.

When the state goes from true to false, a new handler ref captures the currently focused element and the check for open fails and the callback is done. I think what you really want to do is capture the changing state of `popupOpen`:

```suggestion
      if (popupRef.current && popupOpen) {
        const elem = getFirstFocusableElement(popupRef.current);
        if (elem) {
          elem.focus();
        }
      } else if (handlerRef && !popupOpen) {
        handlerRef.focus();
      }
```